### PR TITLE
First version of the Iceberg sink for Apache Beam

### LINKED ISSUE #693: Support for Apache Beam I/O
Is Apache Beam I/O in the roadmap? 

Having Beam I/O can help us integrate Iceberg directly with beam pipelines.

IcebergIO like AvroIO(https://beam.apache.org/documentation/io/built-in/) can help a lot.

### PR DESCRIPTION
**Pull request by Fokko**
The first attempt of adding a sink to write data to a DFS using Beam.

Fixes #693

### CODE REVIEW

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 1]
```diff
@@ -58,9 +58,11 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class HiveCatalog extends BaseMetastoreCatalog implements Closeable, SupportsNamespaces {
+public class HiveCatalog extends BaseMetastoreCatalog implements AutoCloseable, SupportsNamespaces {
```
Comment (rdblue) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:61]: It doesn't look like these changes are needed. Could we move the changes in this file to a separate PR?

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import org.apache.iceberg.hive.HiveCatalog;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+
+import java.util.LinkedList;
+import java.util.List;
+
+
+public class IcebergIO {
+
+    public static void write(TableIdentifier table, Schema schema, String hiveMetastoreUrl, WriteFilesResult<Void> resultFiles) {
```
Comment (rdblue) [inline beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:36]: This doesn't conform to the normal style we use in the project, which is 2 spaces for an indent.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.iceberg.beam;
+
+import org.apache.beam.sdk.io.WriteFilesResult;
+import org.apache.beam.sdk.transforms.Combine;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
```
Comment (rdblue) [inline beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:20]: Can you use the Iceberg version of Guava? Unless you're passing Guava objects to the Beam API, this should use Iceberg's bundled Guava version. (And hopefully there are no places where Beam requires relocated Guava classes!)

**Comment by rdblue** [comment]
@Fokko, thanks for working on this! This looks like it may not be complete. Do you think it should be a draft?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 1 above)
Comment (HeartSaVioR) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:61]: I'm not sure this diff "technically" change anything. Sounds unnecessary unless explained.

Comment (Fokko) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:61]: I was under the impression that we couldn't use `with { } `. But this seems incorrect. I'll revert this change. Thanks for pointing it out, so we can keep the history clean 👍

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java (hunk 3 above)
Comment (Fokko) [reply beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:20]: Auch, great catch!

**Comment by Fokko** [comment]
@rdblue @HeartSaVioR Thanks for the review. This is indeed still a draft. I'll add some more tests soon 👍

**Comment by Fokko** [comment]
Able to read the results with Spark:
```
MacBook-Pro-van-Fokko:incubator-iceberg fokkodriesprong$ spark-sql --packages org.apache.iceberg:iceberg-spark3:0.10.0 --conf "spark.sql.catalog.hive_prod=org.apache.iceberg.spark.SparkCatalog" --conf "spark.sql.catalog.hive_prod.type=hive" --conf "spark.sql.catalog.hive_prod.uri=thrift://localhost:9083"
spark-sql> SELECT * FROM hive_prod.default.test;
20/12/26 22:07:46 INFO BaseMetastoreTableOperations: Refreshing table metadata from new version: file:/tmp/test/metadata/00004-ed964aeb-8924-492b-8225-eb2102c22f03.metadata.json
20/12/26 22:07:46 INFO BaseMetastoreCatalog: Table loaded by catalog: hive_prod.default.test
20/12/26 22:07:47 INFO MemoryStore: Block broadcast_0 stored as values in memory (estimated size 315.8 KiB, free 366.0 MiB)
20/12/26 22:07:48 INFO MemoryStore: Block broadcast_0_piece0 stored as bytes in memory (estimated size 27.4 KiB, free 366.0 MiB)
20/12/26 22:07:48 INFO BlockManagerInfo: Added broadcast_0_piece0 in memory on 192.168.1.113:56438 (size: 27.4 KiB, free: 366.3 MiB)
20/12/26 22:07:48 INFO SparkContext: Created broadcast 0 from broadcast at SparkScanBuilder.java:138
20/12/26 22:07:48 INFO MemoryStore: Block broadcast_1 stored as values in memory (estimated size 40.0 B, free 366.0 MiB)
20/12/26 22:07:48 INFO MemoryStore: Block broadcast_1_piece0 stored as bytes in memory (estimated size 116.0 B, free 366.0 MiB)
20/12/26 22:07:48 INFO BlockManagerInfo: Added broadcast_1_piece0 in memory on 192.168.1.113:56438 (size: 116.0 B, free: 366.3 MiB)
20/12/26 22:07:48 INFO SparkContext: Created broadcast 1 from broadcast at SparkScanBuilder.java:139
20/12/26 22:07:48 INFO V2ScanRelationPushDown: 
Pushing operators to hive_prod.default.test
Pushed Filters: 
Post-Scan Filters: 
Output: word#0
         
20/12/26 22:07:48 INFO TableScan: Scanning table hive_prod.default.test snapshot 4784653135298349855 created at 2020-12-26 22:07:38.530 with filter true
20/12/26 22:07:49 INFO CodeGenerator: Code generated in 226.493784 ms
20/12/26 22:07:49 INFO SparkContext: Starting job: main at NativeMethodAccessorImpl.java:0
20/12/26 22:07:49 INFO DAGScheduler: Got job 0 (main at NativeMethodAccessorImpl.java:0) with 1 output partitions
20/12/26 22:07:49 INFO DAGScheduler: Final stage: ResultStage 0 (main at NativeMethodAccessorImpl.java:0)
20/12/26 22:07:49 INFO DAGScheduler: Parents of final stage: List()
20/12/26 22:07:49 INFO DAGScheduler: Missing parents: List()
20/12/26 22:07:49 INFO DAGScheduler: Submitting ResultStage 0 (MapPartitionsRDD[3] at main at NativeMethodAccessorImpl.java:0), which has no missing parents
20/12/26 22:07:49 INFO MemoryStore: Block broadcast_2 stored as values in memory (estimated size 7.1 KiB, free 366.0 MiB)
20/12/26 22:07:49 INFO MemoryStore: Block broadcast_2_piece0 stored as bytes in memory (estimated size 3.5 KiB, free 366.0 MiB)
20/12/26 22:07:49 INFO BlockManagerInfo: Added broadcast_2_piece0 in memory on 192.168.1.113:56438 (size: 3.5 KiB, free: 366.3 MiB)
20/12/26 22:07:49 INFO SparkContext: Created broadcast 2 from broadcast at DAGScheduler.scala:1223
20/12/26 22:07:49 INFO DAGScheduler: Submitting 1 missing tasks from ResultStage 0 (MapPartitionsRDD[3] at main at NativeMethodAccessorImpl.java:0) (first 15 tasks are for partitions Vector(0))
20/12/26 22:07:49 INFO TaskSchedulerImpl: Adding task set 0.0 with 1 tasks
20/12/26 22:07:49 INFO TaskSetManager: Starting task 0.0 in stage 0.0 (TID 0, 192.168.1.113, executor driver, partition 0, PROCESS_LOCAL, 11598 bytes)
20/12/26 22:07:49 INFO Executor: Running task 0.0 in stage 0.0 (TID 0)
...
20/12/26 22:07:50 INFO Executor: Finished task 0.0 in stage 0.0 (TID 0). 1580 bytes result sent to driver
20/12/26 22:07:50 INFO TaskSetManager: Finished task 0.0 in stage 0.0 (TID 0) in 1132 ms on 192.168.1.113 (executor driver) (1/1)
20/12/26 22:07:50 INFO TaskSchedulerImpl: Removed TaskSet 0.0, whose tasks have all completed, from pool 
20/12/26 22:07:50 INFO DAGScheduler: ResultStage 0 (main at NativeMethodAccessorImpl.java:0) finished in 1,235 s
20/12/26 22:07:50 INFO DAGScheduler: Job 0 is finished. Cancelling potential speculative or zombie tasks for this job
20/12/26 22:07:50 INFO TaskSchedulerImpl: Killing all running tasks in stage 0: Stage finished
20/12/26 22:07:50 INFO DAGScheduler: Job 0 finished: main at NativeMethodAccessorImpl.java:0, took 1,274538 s
hands
yes
my
go
are
typing
words
Time taken: 4.501 seconds, Fetched 7 row(s)
```

File: beam/src/test/java/org/apache/iceberg/beam/IcebergIOTest.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    private static final List<String> SECOND_WIN_WORDS = SENTENCES.subList(4, 6);
+    private static final Duration WINDOW_DURATION = Duration.standardMinutes(1);
+
+    @Rule
+    public final transient TestPipeline pipeline = TestPipeline.create();
+    final String hiveMetastoreUrl = "thrift://localhost:9083/default";
+
+    private static final PipelineOptions options = TestPipeline.testingPipelineOptions();
+
+    private static final String stringSchema = "{\n" +
```
Comment (rdblue) [inline beam/src/test/java/org/apache/iceberg/beam/IcebergIOTest.java:68]: I'm not sure why this would result in readable data files. Iceberg requires either a schema in the Avro file with field IDs or a name mapping in the Iceberg table to get the field IDs from Avro field names. How did you read the table produced by this test?

File: site/docs/beam.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+## Write records
+
+When sinking Avro files to a distributed file system using `FileIO`, we can catch rely the filenames downstream. We append these filenames to an Iceberg table incrementally.
+
+```java
+final String hiveMetastoreUrl = "thrift://localhost:9083/default";
+final FileIO.Write<Void, GenericRecord> avroFileIO = FileIO.<GenericRecord>write()
+        .via(AvroIO.sink(avroSchema))
+        .to("gs://.../../")
+        .withSuffix(".avro");
```
Comment (rdblue) [inline site/docs/beam.md:32]: This would work if the Avro schema passed to the sink were created from the Iceberg table schema because that sets the field IDs correctly. But I think that's not very obvious to users, so I think it would be better to eventually have an Iceberg sink that uses the Iceberg schema directly like Spark and Flink do.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        try {
+          table = catalog.loadTable(this.tableIdentifier);
+        } catch (NoSuchTableException e) {
+          // If it doesn't exist, we just create the table
+          table = catalog.createTable(this.tableIdentifier, schema);
+        }
+
+        // In case the schema has changed
+        if (table.schema() != this.schema) {
+          table.updateSchema().unionByNameWith(this.schema).commit();
```
Comment (rdblue) [inline beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:100]: Users may choose to do this, but I would not recommend it as the default behavior. It is too easy to write bad data to a table and get the union of two schemas instead of having the write rejected. Since we generally try to make Iceberg work like a normal SQL table, I think that this should not be the default.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        Table table;
+        try {
+          table = catalog.loadTable(this.tableIdentifier);
+        } catch (NoSuchTableException e) {
+          // If it doesn't exist, we just create the table
+          table = catalog.createTable(this.tableIdentifier, schema);
+        }
+
+        // In case the schema has changed
+        if (table.schema() != this.schema) {
```
Comment (rdblue) [inline beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:99]: Schemas can be semantically compared using `table.schema().asStruct().equals(schema.asStruct())`.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+          table.updateSchema().unionByNameWith(this.schema).commit();
+        }
+
+        // Append the new files
+        final AppendFiles app = table.newAppend();
+        // We need to get the statistics, not easy to get them through Beam
```
Comment (rdblue) [inline beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:105]: There is a PR to aggregate the stats while writing an Avro file. You would get them for free using an Iceberg-based file writer instead of the generic Avro one. You'd also get correct record counts and file sizes.

File: api/src/main/java/org/apache/iceberg/catalog/Namespace.java [hunk 9]
```diff
@@ -19,13 +19,14 @@
...
 package org.apache.iceberg.catalog;
 
+import java.io.Serializable;
 import java.util.Arrays;
 import org.apache.iceberg.relocated.com.google.common.base.Joiner;
 
 /**
  * A namespace in a {@link Catalog}.
  */
-public class Namespace {
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/catalog/Namespace.java:28]: Are these changes needed?

**Comment by rdblue** [comment]
Looks like a good start, @Fokko! I'd definitely recommend relying a bit more on Iceberg helpers to do things like write data files, but this does demonstrate the basic elements of writing to an Iceberg table.

**Review by Fokko — COMMENTED** [review]
@rdblue Thanks for the comments, appreciate it.

Both the batch and streaming test are running. Only one issue with filenames. By default Beam encodes the window interval into the filename, including the time. This conflicts with the Hadoop file reader, as this doesn't support semicolons `:`.

```
org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 3.0 failed 1 times, most recent failure: Lost task 0.0 in stage 3.0 (TID 3, 192.168.1.113, executor driver): java.lang.IllegalArgumentException: java.net.URISyntaxException: Relative path in absolute URI: .1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro.crc
	at org.apache.hadoop.fs.Path.initialize(Path.java:205)
	at org.apache.hadoop.fs.Path.<init>(Path.java:171)
	at org.apache.hadoop.fs.Path.<init>(Path.java:93)
	at org.apache.hadoop.fs.ChecksumFileSystem.getChecksumFile(ChecksumFileSystem.java:90)
	at org.apache.hadoop.fs.ChecksumFileSystem$ChecksumFSInputChecker.<init>(ChecksumFileSystem.java:145)
	at org.apache.hadoop.fs.ChecksumFileSystem.open(ChecksumFileSystem.java:346)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:769)
	at org.apache.iceberg.hadoop.HadoopInputFile.newStream(HadoopInputFile.java:157)
	at org.apache.iceberg.avro.AvroIterable.newFileReader(AvroIterable.java:101)
	at org.apache.iceberg.avro.AvroIterable.iterator(AvroIterable.java:77)
	at org.apache.iceberg.spark.source.RowDataReader.open(RowDataReader.java:95)
	at org.apache.iceberg.spark.source.BaseDataReader.next(BaseDataReader.java:85)
	at org.apache.spark.sql.execution.datasources.v2.PartitionIterator.hasNext(DataSourceRDD.scala:79)
	at org.apache.spark.sql.execution.datasources.v2.MetricsIterator.hasNext(DataSourceRDD.scala:112)
	at org.apache.spark.InterruptibleIterator.hasNext(InterruptibleIterator.scala:37)
	at scala.collection.Iterator$$anon$10.hasNext(Iterator.scala:458)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$GeneratedIteratorForCodegenStage1.processNext(Unknown Source)
	at org.apache.spark.sql.execution.BufferedRowIterator.hasNext(BufferedRowIterator.java:43)
	at org.apache.spark.sql.execution.WholeStageCodegenExec$$anon$1.hasNext(WholeStageCodegenExec.scala:729)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$getByteArrayRdd$1(SparkPlan.scala:340)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2(RDD.scala:872)
	at org.apache.spark.rdd.RDD.$anonfun$mapPartitionsInternal$2$adapted(RDD.scala:872)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:349)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:313)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:127)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:446)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1377)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:449)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.URISyntaxException: Relative path in absolute URI: .1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro.crc
	at java.net.URI.checkPath(URI.java:1823)
	at java.net.URI.<init>(URI.java:745)
	at org.apache.hadoop.fs.Path.initialize(Path.java:202)
	... 32 more
```

Avro tools:
```
MacBook-Pro-van-Fokko:incubator-iceberg fokkodriesprong$ avro-tools tojson /tmp/fokko/1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro
20/12/30 12:55:18 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Exception in thread "main" java.lang.IllegalArgumentException: java.net.URISyntaxException: Relative path in absolute URI: .1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro.crc
	at org.apache.hadoop.fs.Path.initialize(Path.java:263)
	at org.apache.hadoop.fs.Path.<init>(Path.java:221)
	at org.apache.hadoop.fs.Path.<init>(Path.java:129)
	at org.apache.hadoop.fs.ChecksumFileSystem.getChecksumFile(ChecksumFileSystem.java:94)
	at org.apache.hadoop.fs.ChecksumFileSystem$ChecksumFSInputChecker.<init>(ChecksumFileSystem.java:149)
	at org.apache.hadoop.fs.ChecksumFileSystem.open(ChecksumFileSystem.java:347)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:899)
	at org.apache.avro.tool.Util.openFromFS(Util.java:88)
	at org.apache.avro.tool.Util.fileOrStdin(Util.java:64)
	at org.apache.avro.tool.DataFileReadTool.run(DataFileReadTool.java:87)
	at org.apache.avro.tool.Main.run(Main.java:67)
	at org.apache.avro.tool.Main.main(Main.java:56)
Caused by: java.net.URISyntaxException: Relative path in absolute URI: .1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro.crc
	at java.net.URI.checkPath(URI.java:1823)
	at java.net.URI.<init>(URI.java:745)
	at org.apache.hadoop.fs.Path.initialize(Path.java:260)
	... 11 more
```

I'll add some asserts to the tests to make sure that everything goes well.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java (hunk 8 above)
Comment (rdblue) [reply beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:180]: Iceberg provides common write classes in the core module that we extend for Flink and Spark object models, so if you do go with Iceberg writers, it would probably not be too much code. For example, take a look at the [`SparkPartitionedWriter`](https://github.com/apache/iceberg/blob/master/spark/src/main/java/org/apache/iceberg/spark/source/SparkPartitionedWriter.java).

Using Iceberg writers would help with stats as well as writing delete files if you later want to support CDC or other row-level use cases.

But, one challenge is integrating those writers with an object model. Spark and Flink have row classes that are directly supported in the integration. It looks like you're using Avro generics for Avro, and using an object model specific to a data format is going to be a little difficult because Iceberg needs additional metadata, like field IDs. It sounds difficult to make sure that each format writer is producing the right information for Iceberg tables.

File: site/docs/beam.md (hunk 5 above)
Comment (rdblue) [reply site/docs/beam.md:32]: The problem with converting Avro to Iceberg is that there is no guarantee that the field IDs are assigned consistently (see [the Field IDs section for Avro](http://iceberg.apache.org/spec/#avro)). If the ID annotations aren't present, they are reassigned but that depends on the structure of the schema.

To safely write Avro for an Iceberg table, you should always convert from the table schema to an Avro schema so that the IDs are correct and match the table.

File: beam/src/test/java/org/apache/iceberg/beam/IcebergIOTest.java (hunk 4 above)
Comment (rdblue) [reply beam/src/test/java/org/apache/iceberg/beam/IcebergIOTest.java]: Okay, so the initial conversion to Iceberg assigns some IDs, then you pass that Iceberg schema into the source. The IDs are reassigned to ensure consistency when a table is created, and you're probably converting that table schema back to Iceberg to configure the Avro writer?

File: api/src/main/java/org/apache/iceberg/catalog/Namespace.java (hunk 9 above)
Comment (rdblue) [reply api/src/main/java/org/apache/iceberg/catalog/Namespace.java:28]: Sounds good. I have no problem making these serializable.

**Comment by rdblue** [comment]
For the file issue, what is the full location that you're passing into Iceberg? Are you sure it isn't missing `file:/` to signal that it is the local FS? I think maybe that's the problem.

**Comment by Fokko** [comment]
@rdblue That was my first thought as well. Unfortunally that's not the case:
```
MacBook-Pro-van-Fokko:dbt-data-ai-summit fokkodriesprong$ avro-tools tojson file:/tmp/fokko/1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro
20/12/30 21:05:11 WARN util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Exception in thread "main" java.lang.IllegalArgumentException: java.net.URISyntaxException: Relative path in absolute URI: .1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro.crc
	at org.apache.hadoop.fs.Path.initialize(Path.java:263)
	at org.apache.hadoop.fs.Path.<init>(Path.java:221)
	at org.apache.hadoop.fs.Path.<init>(Path.java:129)
	at org.apache.hadoop.fs.ChecksumFileSystem.getChecksumFile(ChecksumFileSystem.java:94)
	at org.apache.hadoop.fs.ChecksumFileSystem$ChecksumFSInputChecker.<init>(ChecksumFileSystem.java:149)
	at org.apache.hadoop.fs.ChecksumFileSystem.open(ChecksumFileSystem.java:347)
	at org.apache.hadoop.fs.FileSystem.open(FileSystem.java:899)
	at org.apache.avro.tool.Util.openFromFS(Util.java:88)
	at org.apache.avro.tool.Util.fileOrStdin(Util.java:64)
	at org.apache.avro.tool.DataFileReadTool.run(DataFileReadTool.java:87)
	at org.apache.avro.tool.Main.run(Main.java:67)
	at org.apache.avro.tool.Main.main(Main.java:56)
Caused by: java.net.URISyntaxException: Relative path in absolute URI: .1970-01-01T00-00-00.000+00:00-1970-01-01T00-01-00.000+00:00-00000-of-00001.avro.crc
	at java.net.URI.checkPath(URI.java:1823)
	at java.net.URI.<init>(URI.java:745)
	at org.apache.hadoop.fs.Path.initialize(Path.java:260)
	... 11 more
```
I've added a new Naming Policy to the PR that resolves the issue. Besides that, I've added an assert that catches the semicolon, to make sure that you don't write data, that you can't read using Spark/Hive/etc.

**Comment by rdblue** [comment]
Here's the relevant check in URI that is failing:

```java
            if ((path != null)
                && ((path.length() > 0) && (path.charAt(0) != '/')))
                throw new URISyntaxException(s,
                                             "Relative path in absolute URI");
```

Looks like the location is parsed in `Path`:
```java
    // parse uri scheme, if any
    int colon = pathString.indexOf(':');
    int slash = pathString.indexOf('/');
    if ((colon != -1) &&
        ((slash == -1) || (colon < slash))) {     // has a scheme
      scheme = pathString.substring(0, colon);
      start = colon+1;
    }
```

String's `indexOf` returns on the first occurrence of `:`, so it looks like it must be that the `file:` is getting stripped out somehow. Can you debug and see what's actually passed?

**Comment by Fokko** [comment]
@rdblue Happy new year! :) Looked into this, it was much more interesting (and time consuming :-) than I've anticipated. You were close. Allow me to summarize:

Also good to know, is that Beam does not output the crc files for the Avro data files, and therefore it silently ignores them when it comes to the names without the semicolons:
image

However, with the semicolons, it fails at line 145. So it is unable to open the file. It indeed takes the part before as the scheme:
image

At `Path(Path parent, String child)` in `Path.java` the parent is `/tmp/test_streaming_crc`, and the child `.output-1970-01-01T00:00:00.000Z-1970-01-01T00:01:00.000Z-00000-of-00001.avro.crc`. Which is being split by the Path constructor. It detects the first semicolon, and will proceed if there is no slash. So this will make `.output-1970-01-01T00` the schema, and `00:00.000Z-1970-01-01T00:01:00.000Z-00000-of-00001.avro.crc` the file:

image

However, Java doesn't not accept the semicolons. It boils down to this:
```java
import java.net.URI;

// This works:
new URI(null, null, normalizePath(null, ".00004-e82668ab-25ec-4190-864e-e4d140eb83b5.metadata.json.crc"), null, null);

// This breaks:
new URI(null, null, normalizePath(null, ".output-1970-01-01T00:00:00.000Z-1970-01-01T00:01:00.000Z-00000-of-00001.avro.crc"), null, null);
```
Throwing the Relative path in absolute URI error. This is still an open issue:

- https://issues.apache.org/jira/browse/HADOOP-3257
- https://issues.apache.org/jira/browse/HADOOP-7945
- https://issues.apache.org/jira/browse/HADOOP-12455
- https://issues.apache.org/jira/browse/HADOOP-14217

I'll add a note to the new naming policy, referring to this issue, and leave this as is for now.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java (hunk 8 above)
Comment (Fokko) [reply beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:180]: Avro is a first-class citizen in Beam, this is more or less the equivalent of the InternalRow of Spark. We could replace the AvroIO with an IcebergIO, but this feels like reinventing the wheel to me. Then we still need to current logic of doing the commits. So, let's keep this option open for now.

**Comment by Fokko** [comment]
Works like a charm:
image

**Comment by Fokko** [comment]
How do we feel about this PR? Only Avro write support enough to get this in, or should we also include reading Iceberg? That would take some more time.

**Comment by rdblue** [comment]
@Fokko, sorry for the delay getting back to this. I've been trying to get everything reviewed for the 0.11.0 release.

I'm excited to see Beam support working, but I don't think that this PR has enough guardrails for general use. Because users create their own Avro files, it would be easy to create files without field IDs and append them to a table. Doing that also requires an extra step to read the Avro file if I'm reading `FilenameToDataFile` correctly. We usually want engines to handle the details of writing data correctly so users don't need to understand them, but this PR delegates the problem to users.

I think we would need to provide more support so that users don't write bad data into their tables, which would also fix the performance problem introduced by `FilenameToDataFile`. There is already support in Iceberg for writing Avro generics to both Avro and Parquet formats. Could this add file writers using that to ensure that file stats and schema are correct? That would easily add Parquet support in the process.

**Comment by Fokko** [comment]
@rdblue I agree, first writing with Beam, and computing statistics is very wasteful. Also, Beam isn't native to object stores (some renaming going on), so I'd have to fix that in the Beam codebase as well.

Got the Iceberg writers working today. Took less code than I anticipated. It is still very rough, so please don't look at the code :) Will clean this up tomorrow and add tests! Thanks!

**Comment by iemejia** [comment]
@Fokko on the interest of making Beam more object store friendly don't hesitate to bring ideas/discussion into the Beam community. I think we need to improve in this area. Does this have an associated JIRA in the Iceberg side? I want to link it to [BEAM-10160](https://issues.apache.org/jira/browse/BEAM-10160)

**Comment by Fokko** [comment]
@iemejia I would like to sort this out, before opening a new discussion :) But I think it is an import thing to address, as it slows down the throughput of the pipeline, and increases the costs (because of the increased number of operations). Iceberg doesn't rely on renames, as it is build with object stores in mind.

I'm happy to add the read part later on, but if you feel like it, feel free to pick it up. I got quite some backlog.

**Comment by iemejia** [comment]
Please go ahead with the Read @Fokko I was just curious. I do have just too much in my backlog too to even risk to add more stuff. Let's sync for the Read part because I have a [WIP branch for Delta](https://github.com/iemejia/beam/tree/deltaio) and I suppose both can end up being similar (ok saying this with my poor knowledge of both Delta and Iceberg).

We can also discuss once you have finished the write (and/or read) on how to integrate it with Beam's SQL, it is quite straight-forward.

**Comment by aokolnychyi** [comment]
Thanks for the great work, @Fokko! 

We have a person who is also interested in a Beam sink for Iceberg. It would be great to collaborate. @RussellSpitzer and I can help with reviews. I feel like this is not far from being ready.

A couple of questions from me (I have just a basic understanding of Beam):
- Is it correct that this PR supports only fixed windows in Beam? 
- I head it is pretty common to write to multiple tables or determine which table to write to dynamically in Beam. Is it worth supporting such cases in the Iceberg sink? Does not have to be done in the first version, apparently.
- Does Beam have any way to control the distribution and ordering of data that is passed for write? For example, we have added a way to ask Spark to distribute and order records according to the partition spec and sort order. A similar effort is in progress for Flink. The main reason is to reduce the number of produced files and order records accordingly to benefit from min/max skipping and better compression. Can people shuffle and sort data manually? Can we do that automatically?


The first two points are minor but the last one can be a big deal at scale. For example, if we don't distribute and order data on write, we either need to produce a lot of small files or keep a lot of files open.

File: beam/src/main/java/org/apache/iceberg/beam/FileWriter.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+          @Override
+          protected PartitionKey partition(T row) {
+            return new PartitionKey(spec, schema);
+          }
+        };
+      }
+    }
+    try {
+      lastSeenWindow = window;
+      writer.write(element);
```
Comment (aokolnychyi) [inline beam/src/main/java/org/apache/iceberg/beam/FileWriter.java:158]: What if we get records for multiple partitions? I think `PartitionedWriter` will fail with a runtime exception.

**Comment by aokolnychyi** [comment]
@madhubabubits, let us know if you have any other comments.

File: beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+        .apply(
+            "Write DataFiles",
+            ParDo.of(new FileWriter<>(table, schema, PartitionSpec.unpartitioned(), hiveMetastoreUrl, properties))
+        )
+        .setCoder(SerializableCoder.of(DataFile.class));
+
+    // We use a combiner, to combine all the files to a single commit in
+    // the Iceberg log
+    final IcebergDataFileCommitter combiner = new IcebergDataFileCommitter(table, hiveMetastoreUrl, properties);
+    final Combine.Globally<DataFile, Snapshot> combined = Combine.globally(combiner).withoutDefaults();
```
Comment (madhubabubits) [inline beam/src/main/java/org/apache/iceberg/beam/IcebergIO.java:97]: @Fokko - firstly, great work with the implementation.

Do you see any use for returning the snapshot here ? Instead, can we implement something on the lines of FileIO sink ? that way user will have full control on windowing, dynamic tablename ( based on the avro schema of each GenericRecord in PCollection<T extends GenericRecord> ) and support both bounded & unbounded collection. 

Also, do you see value in providing windowing for both creating files & windowing for committing ( like create files based on 5min window, but final commit is based on 1hr window ). This might be a use-case for unbounded collection

File: api/src/main/java/org/apache/iceberg/DataFile.java [hunk 12]
```diff
@@ -35,7 +36,7 @@
 /**
  * Interface for data files listed in a table manifest.
  */
-public interface DataFile extends ContentFile<DataFile> {
+public interface DataFile extends ContentFile<DataFile>, Serializable {
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/DataFile.java:39]: If we are implementing Serializable, should we add serial version uid?

In particular, I have users that run into issues with things that get stored in state having changed version uids when they're not explicitly pinned.

File: api/src/main/java/org/apache/iceberg/catalog/TableIdentifier.java [hunk 13]
```diff
@@ -28,7 +29,7 @@
 /**
  * Identifies a table in iceberg catalog.
  */
-public class TableIdentifier {
+public class TableIdentifier implements Serializable {
```
Comment (kbendick) [inline api/src/main/java/org/apache/iceberg/catalog/TableIdentifier.java:32]: For some of these classes where they are implementing Serializable, should we consider adding a `serialVersionUID`?

I have many users (specifically for Flink) that encounter subtle issues when some Serializable class does not specify `serialVersionUID` and the compiler changes the assigned seriallVersionUID. Very often, these users need to allow some or all of their jobs existing state to be discarded when this happens.

I'm not sure if `TableIdentifier` gets placed into state anywhere ourselves, but for classes that are part of the public api is there an argument against including them?

Additionally, does anybody have any arguments for or against including serial version uid on the interfaces which are also extending serializable? For example, `DataFile` stands out to me in particular as things that could potentially cause issue. I could imagine that something along the lines of `Class<DataFile> clazz = DataFile.class`  could potentially get serialized and cause issues when the interface changes.

**Comment by Fokko** [comment]
Bit out of date :)

**Comment by sachinag** [comment]
To the extent we on the Beam team can help, please let me (us) know.  We'd love to see this completed and merged.
